app.rb

```diff
@@ -0,0 +1,185 @@
+require 'sinatra'
+require "sinatra/reloader"
+require 'sinatra/base'
+require 'active_record'
+require 'pg'
+require 'cgi'
+require_relative 'lib/peep'
+require_relative 'lib/user'
+require_relative 'lib/send_email'
+
+ActiveRecord::Base.establish_connection(
+  adapter:  'postgresql',
+  database: 'chitter'
+)
+
+class Application < Sinatra::Base
+
+  configure :development do
+    register Sinatra::Reloader
+    also_reload 'lib/peep'
+    also_reload 'lib/user'
+    enable :sessions
+  end
+
+  def self.escape_html_all_params(params)
+    # this santises all inputs
+    # included as a self method so it could be tested in RSpec
+    params.each { |k, v| params[k] = CGI.escapeHTML(v) }
+  end
+
+  get '/' do
+    # loads the homepage
+
+    # :user_id == nil as default, when a user logs in this changes to their User.id
+    # :user_id.nil? is how we check whether a user is logged in
+    # :user_id == nil => login => :user_id == 1
+    @user_id = session[:user_id]
+    # loads the username for the page if logged in
+    @username = @user_id.nil? ? nil : User.find(@user_id).username
+    # loads the last 20 Peeps to a timeline
+    @peeps = Peep.last(20).sort_by { |peep| peep.date_time_created }.reverse
+    erb(:index)
+  end
+
+  get '/login' do
+    # loads the login form page
+
+    # redirects back to the homepage if user is logged in already
+    redirect to '/' unless session[:user_id].nil?
+    # passing into variables for erb
+    @user_id = session[:user_id]
+    @sign_up = params[:sign_up] == 'true'
+    # loads error messages in erb if needed
+    @username_error = params[:username_error] == 'true'
+    @password_error = params[:password_error] == 'true'
+    erb(:login)
+  end
+
+  post '/login' do
+    # verifies login info
+
+    # returns error status 400 if no user is logged in or parameters are given
+    return status 400 if !session[:user_id].nil? ||
+      params[:password].nil? || params[:username].nil?
+    Application.escape_html_all_params(params)
+
+    # checks username is in database, nil if not
+    user = User.find_by(username: params[:username])
+
+    # redirect back to login page with username error message if user == nil
+    redirect to '/login?username_error=true' if user.nil?
+
+    # redirect back to login page with password error message if password doesn't match
+    redirect to '/login?password_error=true' if params[:password] != user.password
+    session[:user_id] = user.id
+    redirect to '/'
+  end
+
+  get '/log-off' do
+    # logs the user off
+
+    # resets :user_id to nil
+    session.delete(:user_id)
```

Comment:
you might need to delete all the session values u created , like id and password